Bind DHCP received DNS server to correct network interface

jukkar opened the pull request:
When we receive DNS servers via DHCPv4/6 options, bind the servers to certain network interface. This can be controlled by a Kconfig option if that is not desired. Add also support for removing the DNS servers from a network interface when that interface goes down.

Fixes #88915

jukkar committed:
net: dns: Allow reconfiguration with network interfaces

Allow user to specify which network interfaces to bind the
server to when reconfiguring the DNS system.

Signed-off-by: Jukka Rissanen <jukka.rissanen@nordicsemi.no>

jukkar committed:
net: dhcpv4: Honor network interface for DNS servers

Bind the network interface to the network interface we
have received the DNS servers from. This is now the default.
The previous behavior can be restored by disabling the
CONFIG_NET_DHCPV4_DNS_SERVER_VIA_INTERFACE option.

Signed-off-by: Jukka Rissanen <jukka.rissanen@nordicsemi.no>

jukkar committed:
net: dhcpv6: Honor network interface for DNS servers

Bind the network interface to the network interface we
have received the DNS servers from. This is now the default.
The previous behavior can be restored by disabling the
CONFIG_NET_DHCPV6_DNS_SERVER_VIA_INTERFACE option.

Signed-off-by: Jukka Rissanen <jukka.rissanen@nordicsemi.no>

jukkar committed:
net: dns: Properly cleanup DNS servers per network interface

Make sure we cleanup only those DNS servers that belong to
certain network interface when the interface goes down.

Signed-off-by: Jukka Rissanen <jukka.rissanen@nordicsemi.no>

jukkar committed:
net: shell: conn: Print packet socket information correctly

The "net conn" command printed AF_PACKET socket as AF_UNK(3).
Fix this by printing "AF_PACKET" in this case.

Signed-off-by: Jukka Rissanen <jukka.rissanen@nordicsemi.no>

jukkar committed:
net: dns: Add uninitialized state to resolver context

Make sure that uninitialized state is 0 so that we can catch
when the context is not initialized properly.

Signed-off-by: Jukka Rissanen <jukka.rissanen@nordicsemi.no>

jukkar committed:
drivers: wifi: esp_at: Bind DNS to device net interface

Make sure received DNS server information is bound to the
wifi network interface.

Signed-off-by: Jukka Rissanen <jukka.rissanen@nordicsemi.no>

rlubos commented on include/zephyr/net/dns_resolve.h:
```diff
@@ -524,6 +524,32 @@ int dns_resolve_reconfigure(struct dns_resolve_context *ctx,
+ * @param servers_sa DNS server addresses as struct sockaddr. The array
+ * is NULL terminated. Port numbers are optional in struct sockaddr, the
+ * default will be used if set to 0.
+ * @param interfaces Network interfaces to which the DNS servers are bound.
+ *        This is an array of network interface indices. The array must be
+ *        the same length as the servers_str and servers_sa arrays.
```
nit: This entry is aligned differently than others (and TBH I prefer this one, it's easier to read if indented).

rlubos commented on subsys/net/lib/dhcpv4/dhcpv4.c:
```diff
@@ -1668,6 +1668,16 @@ static void dhcpv4_iface_event_handler(struct net_mgmt_event_callback *cb,
+			 * DNS servers are removed automatically when the interface
+			 * comes back up.
+			 */
+			if (IS_ENABLED(CONFIG_NET_DHCPV4_DNS_SERVER_VIA_INTERFACE)) {
```
Shouldn't this only be done if `CONFIG_NET_DHCPV4_OPTION_DNS_ADDRESS` is enabled? I imagine otherwise DNS addresses could be statically configured.

jukkar replied:
It is done only if `CONFIG_NET_DHCPV4_OPTION_DNS_ADDRESS` is enabled (see line 1120 above where this is enforced). Also there is a dependency in Kconfig for these options.

rlubos replied:
Ah true, you're right, nevermind then.

rlubos commented on subsys/net/lib/dhcpv6/dhcpv6.c:
```diff
@@ -2217,6 +2217,16 @@ static void dhcpv6_iface_event_handler(struct net_mgmt_event_callback *cb,
+		 * DNS servers are removed automatically when the interface
+		 * comes back up.
+		 */
+		if (IS_ENABLED(CONFIG_NET_DHCPV6_DNS_SERVER_VIA_INTERFACE)) {
```
Ditto, but for DHCPV6 option.

jukkar replied:
There is a dependency in Kconfig that will enforce this.

rlubos reviewed (commented):
Nice! Not sure if we even need the Kconfig for the new behaviour, as it's quite sane, but it shouldn't hurt to have it.

jukkar commented:
* updated according to comments

rlubos reviewed (dismissed):
LGTM

jukkar committed:
net: ppp: Bind DNS to device ppp network interface

Make sure received DNS server information is bound to the
ppp network interface.

Signed-off-by: Jukka Rissanen <jukka.rissanen@nordicsemi.no>

jukkar commented:
* PPP was missing DNS server runtime binding
